File: Python/백준/Gold/31531. 공들의 리듬게임/공들의 리듬게임.py
N,*A=open(0)
N=int(N)
A=[[*map(int,i.split())]for i in A]
A.sort(key=lambda x:x[0])
L=R=ans=0
for a,d in A:
    if d==1:R+=1
    elif d==0:ans+=R*2
    else:ans+=R
for a,d in A[::-1]:
    if d==-1:L+=1
    elif d==0:ans+=L*2
print(ans)

File: Java/백준/Gold/31531. 공들의 리듬게임/공들의 리듬게임.java
import java.io.*;
import java.util.*;

public class Main {
    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static StringTokenizer st;
    static StringBuilder sb = new StringBuilder();

    static int N, ans, right, left;
    static ArrayList<int[]> arr;

    public static void main(String[] args) throws IOException {
        N = Integer.parseInt(br.readLine());
        arr = new ArrayList<>();
        for (int i = 0; i < N; i++) {
            st = new StringTokenizer(br.readLine());
            arr.add(new int[] {Integer.parseInt(st.nextToken()), Integer.parseInt(st.nextToken())});
        }
        arr.sort((o1, o2) -> o1[0] - o2[0]);
        for (int[] is : arr) {
            int a = is[0];
            int d = is[1];
            if (d == 1) right++;
            else if (d == 0) ans += right*2;
            else ans += right;
        }
        
        Collections.reverse(arr);
        for (int[] is : arr) {
            int a = is[0];
            int d = is[1];
            if (d == -1) left++;
            else if (d == 0) ans += left*2;
        }
        System.out.println(ans);
    }
}